Python/ceval.c

```diff
@@ -5221,19 +5221,28 @@ import_from(PyThreadState *tstate, PyObject *v, PyObject *name)
         pkgname_or_unknown = pkgname;
     }
 
+    _Py_IDENTIFIER(__spec__);
+    PyObject* spec = _PyObject_GetAttrId(v, &PyId___spec__);
+    Py_XINCREF(spec);
+    const char* partially_initialized =
+        _PyModuleSpec_IsInitializing(spec) ?
+        " (partially initialized module -- most likely due to a circular import)" :
+        "";
+    Py_XDECREF(spec);
+
     if (pkgpath == NULL || !PyUnicode_Check(pkgpath)) {
         _PyErr_Clear(tstate);
         errmsg = PyUnicode_FromFormat(
-            "cannot import name %R from %R (unknown location)",
-            name, pkgname_or_unknown
+            "cannot import name %R from %R%s (unknown location)",
```

Comment:
This case is not tested.

Reply:
indeed, I couldn't figure out how to test the "unknown location" errors

Reply:
Change the type of the exception raised here, run tests, and see what tests are failed. After that you can modify the affected tests or add similar tests which check also error message.

Reply:
Seems to only happen for `ImportError` from builtin modules so it should be impossible -- I'll factor that out 👍